base64_encode.go

```diff
@@ -0,0 +1,47 @@
+package gophplib
+
+import (
+	"encoding/base64"
+	"fmt"
+	"math"
+	"reflect"
+)
+
+// Base64Encode emulates the functionality of PHP 5.6's base64_encode function.
+// For more information, see the [official PHP documentation].
+// In PHP 5.6, an error is triggered if the input size is excessively large due to memory limitations.
+// This Go implementation includes similar checks to emulate PHP's memory limitation conditions.
+// Additionally, this function converts different types of variables to a string, following PHP's dynamic typing approach.
+// After ensuring the memory constraints are met and converting the input to a string,
+// it uses the EncodeToString function from the encoding/base64 package to perform the Base64 encoding.
+// The result is a Base64 encoded string, consistent with PHP's output for the same input.
+// For more detailed information about the EncodeToString function in the package encoding/base64,
+// see the [encoding/base64's EncodeToString documentation]
+//
+// PHP references:
+//   - base64_encode definition:
+//     https://github.com/php/php-src/blob/php-5.6.40/ext/standard/base64.c#L224-L241
+//   - base64_encode implementation:
+//     https://github.com/php/php-src/blob/4b8f72da5dfb201af4e82dee960261d8657e414f/ext/standard/base64.c#L56-L106
+//
+// Test Cases :
+//   - https://github.com/php/php-src/blob/php-5.6.40/ext/standard/tests/url/base64_encode_basic_001.phpt
+//   - https://github.com/php/php-src/blob/php-5.6.40/ext/standard/tests/url/base64_encode_basic_002.phpt
+//   - https://github.com/php/php-src/blob/php-5.6.40/ext/standard/tests/url/base64_encode_error_001.phpt
+//   - https://github.com/php/php-src/blob/php-5.6.40/ext/standard/tests/url/base64_encode_variation_001.phpt
+//
+// [official PHP documentation]: https://www.php.net/manual/en/function.base64-encode
+// [encoding/base64's EncodeToString documentation]: https://pkg.go.dev/encoding/base64#Encoding.EncodeToString
+func Base64Encode(value any) (string, error) {
+	// Convert a value to string
+	characterString, err := zendParseArgAsString(value)
+	if err != nil {
+		return "", fmt.Errorf("unsupported type : %s", reflect.TypeOf(value))
+	}
+
+	if (len(characterString)+2)/3 > math.MaxInt32/4 {
```

Comment:
Given Go's efficient memory management and performance capabilities, I conducted a benchmark test to evaluate if this limitation is indeed necessary.

```go
func BenchmarkBase64Encode(b *testing.B) {
    for _, size := range []int{100, 10000, 1000000, 100000000, 10000000000} {
        b.Run(fmt.Sprintf("Size%d", size), func(b *testing.B) {
            testString := bytes.Repeat([]byte("a"), size)
            b.ResetTimer()
            for i := 0; i < b.N; i++ {
                _, err := Base64Encode(string(testString))
                if err != nil {
                    b.Fatalf("error: %v", err)
                }
            }
        })
    }
}
```

```bash
go test -benchmem -run=^$ -bench ^BenchmarkBase64Encode$ github.com/portone-io/gophplib
goos: darwin
goarch: arm64
pkg: github.com/portone-io/gophplib
BenchmarkBase64Encode/Size100-8                  6569655               189.9 ns/op           528 B/op          5 allocs/op
BenchmarkBase64Encode/Size10000-8                 124491              9642 ns/op           47632 B/op          5 allocs/op
BenchmarkBase64Encode/Size1000000-8                 1522            825095 ns/op         4685846 B/op          5 allocs/op
BenchmarkBase64Encode/Size100000000-8                 15          67598669 ns/op        466698275 B/op         5 allocs/op
BenchmarkBase64Encode/Size536870911-8                  1        1350157917 ns/op        2505408720 B/op        7 allocs/op
PASS
ok      github.com/portone-io/gophplib  8.382s
```

The benchmark clearly shows that the function can handle strings up to 536,870,911 bytes (approximately 536.87 MB) efficiently. The linear increase in time and memory allocation with the increase in string size demonstrates Go's capability to handle large strings effectively.

So, In my view, managing large inputs and potential memory-related errors at a higher level (such as the application layer which directly utilizes this function) could be more effective than incorporating restrictive logic within the function itself.

Reply:
I agree that managing memory at the application layer is more efficient as it allows for larger strings to be encoded, but this function aims to exactly match the behaviour in PHP5.6, and since PHP's base64_encode function set memory limits on implementations, this function is also memory-limited.

Reply:
I agree with keeping this restriction in gophplib as well. It's a more predictable, more boring approach, and probably the behavior that gophplib's users want.

Reply:
I understand and will close this issue.